File: scripts/measure-boot-timing.ts
#!/usr/bin/env bun
/**
 * measure-boot-timing.ts
 *
 * Measure empty-script engine-boot wall-clock for GocciaScriptLoader in both
 * execution modes and emit a small JSON summary the PR suite-timing comment
 * renders.
 *
 * Boot wall-clock on a shared runner is dominated by a fixed process /
 * dynamic-loader floor, so we report the min (least-contended run) alongside
 * the median over many runs rather than a single timing. The two modes are
 * measured back-to-back so transient load affects both alike.
 *
 * Run: bun run scripts/measure-boot-timing.ts <loader> <output.json>
 * Env: BOOT_TIMING_RUNS (default 60), BOOT_TIMING_WARMUP (default 10)
 */

import { mkdtempSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

const [, , loader, output] = Bun.argv;
if (!loader || !output) {
  console.error("usage: bun run scripts/measure-boot-timing.ts <loader> <output.json>");
  process.exit(2);
}

function intEnv(name: string, fallback: number, min: number): number {
  const raw = Bun.env[name];
  if (raw === undefined || raw === "") return fallback;
  const value = Number(raw);
  if (!Number.isInteger(value) || value < min) {
    throw new Error(`${name} must be an integer >= ${min}, got: ${raw}`);
  }
  return value;
}

const runs = intEnv("BOOT_TIMING_RUNS", 60, 1);
const warmup = intEnv("BOOT_TIMING_WARMUP", 10, 0);

interface ModeTiming {
  minMs: number;
  medianMs: number;
}

function measure(script: string, modeArgs: string[]): ModeTiming {
  const samples: number[] = [];
  const label = modeArgs.join(" ") || "interpreter";
  for (let i = 0; i < warmup + runs; i++) {
    const start = Bun.nanoseconds();
    const result = Bun.spawnSync([loader, script, ...modeArgs], {
      stdout: "ignore",
      stderr: "ignore",
    });
    const elapsedMs = (Bun.nanoseconds() - start) / 1e6;
    // A crashed/non-zero child run is not a valid boot sample — fail rather than
    // write boot-timing.json from bogus timings.
    if (result.exitCode !== 0) {
      throw new Error(`loader exited ${result.exitCode} measuring boot (${label})`);
    }
    if (i >= warmup) samples.push(elapsedMs);
  }
  samples.sort((a, b) => a - b);
  const n = samples.length;
  const median =
    n % 2 ? samples[(n - 1) / 2] : (samples[n / 2 - 1] + samples[n / 2]) / 2;
  return {
    minMs: Math.round(samples[0] * 100) / 100,
    medianMs: Math.round(median * 100) / 100,
  };
}

const dir = mkdtempSync(join(tmpdir(), "goccia-boot-"));
try {
  const script = join(dir, "empty.js");
  writeFileSync(script, "0;\n");

  const report = {
    runs,
    warmup,
    interpreted: measure(script, []),
    bytecode: measure(script, ["--mode=bytecode"]),
  };

  const json = JSON.stringify(report, null, 2);
  writeFileSync(output, json + "\n");
  console.log(json);
} finally {
  rmSync(dir, { recursive: true, force: true });
}

File: .github/scripts/measure-boot-timing.py
#!/usr/bin/env python3
"""Measure empty-script engine-boot wall-clock for GocciaScriptLoader in both
execution modes and emit a small JSON summary the PR suite-timing comment
renders.

Boot wall-clock on a shared runner is dominated by a fixed process /
dynamic-loader floor, so we report the min (least-contended run) alongside the
median over many runs rather than a single timing. The two binaries' modes are
measured back-to-back so transient load affects both alike.

Usage: measure-boot-timing.py <loader> <output.json>
Env:   BOOT_TIMING_RUNS (default 60), BOOT_TIMING_WARMUP (default 10)
"""
import json
import os
import statistics
import subprocess
import sys
import tempfile
import time


def measure(loader: str, script: str, mode_args: list[str], runs: int, warmup: int) -> dict:
    samples = []
    for i in range(warmup + runs):
        start = time.perf_counter()
        subprocess.run(
            [loader, script, *mode_args],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
        )
        elapsed_ms = (time.perf_counter() - start) * 1000.0
        if i >= warmup:
            samples.append(elapsed_ms)
    return {
        "minMs": round(min(samples), 2),
        "medianMs": round(statistics.median(samples), 2),
    }


def main() -> int:
    if len(sys.argv) != 3:
        print("usage: measure-boot-timing.py <loader> <output.json>", file=sys.stderr)
        return 2
    loader, out = sys.argv[1], sys.argv[2]
    runs = int(os.environ.get("BOOT_TIMING_RUNS", "60"))
    warmup = int(os.environ.get("BOOT_TIMING_WARMUP", "10"))

    with tempfile.TemporaryDirectory() as tmp:
        script = os.path.join(tmp, "empty.js")
        with open(script, "w", encoding="utf-8") as handle:
            handle.write("0;\n")

        report = {
            "runs": runs,
            "warmup": warmup,
            "interpreted": measure(loader, script, [], runs, warmup),
            "bytecode": measure(loader, script, ["--mode=bytecode"], runs, warmup),
        }

    with open(out, "w", encoding="utf-8") as handle:
        json.dump(report, handle, indent=2)
        handle.write("\n")
    print(json.dumps(report, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
